modules/imgproc/src/peronamalik.cpp

```diff
@@ -0,0 +1,89 @@
+
+#include "precomp.hpp"
+#include "opencl_kernels_imgproc.hpp"
+#include "cvconfig.h"
+
+#include <iostream>
+using namespace std;
+
+namespace cv
+{
+    void PeronaMalik(InputArray src, OutputArray dst, double timeStepSize, double k, int noOfTimeSteps, int fluxFunc )
+    {
+        CV_Assert(k != 0);
+        CV_Assert(timeStepSize > 0);
+        CV_Assert(noOfTimeSteps >= 0);
+        CV_Assert((fluxFunc == CV_PERONA_MALIK_EXPONENTIAL) || (fluxFunc == CV_PERONA_MALIK_INVERSE_QUADRATIC));
+        CV_Assert(src.size() == dst.size());
+        CV_Assert(src.type() == dst.type());
+        
+        Mat _dst = dst.getMat();
+        Mat _src = src.getMat().clone();
+        
+        int channels = _src.channels();
+        int nRows = _src.rows;
+        int nCols = _src.cols * channels;
+        
+        double oneOnkSquared = 1 / (k * k);
+        
+        int i,j;
+        for (int t = 0; t < noOfTimeSteps; t++)
+        {
+            for( i = 0; i < nRows; ++i)
+            {
+                double* p_src_S = _src.ptr<double>(min(i+1, nRows-1));
+                double* p_src   = _src.ptr<double>(i);
+                double* p_src_N = _src.ptr<double>(max(i-1, 0));
+                double* p_dst   = _dst.ptr<double>(i);
+                
+                for ( j = 0; j < nCols; ++j)
+                {
+                    int ch = j % channels;
+                    double center  = p_src[j];
+                    double nablaN  = p_src_N[j]                         - center; 
+                    double nablaS  = p_src_S[j]                         - center;
+                    double nablaE  = p_src[min(j+channels,   nCols-ch)] - center;
```

Comment:
because the processing is quite heavy, you may consider to add a border to the image (using copyMakeBorder) and then all those min and max operations can be eliminated